### PR TITLE
NXP-30171: return http status code based on exception thrown

### PR DESCRIPTION
**Pull request by efge**

### CODE REVIEW

File: modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoRequestControllerFilter.java [hunk 1]
```diff
@@ -138,9 +138,18 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
             }
             chain.doFilter(request, response);
         } catch (IOException | ServletException | RuntimeException e) {
+            if (!response.isCommitted()) {
+                response.resetBuffer();
```
Comment (nelsonsilva) [inline modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoRequestControllerFilter.java:142]: this change is not directly related, right?

File: modules/platform/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/impl/OperationChainCompiler.java [hunk 2]
```diff
@@ -214,9 +214,12 @@ public Object invoke(OperationContext ctx) throws OperationException {
                     if (e.isRollback()) {
                         ctx.setRollback();
                     }
-                    return ctx.getInput();
-                } catch (OperationException op) {
-                    throw ctx.getCallback().onError(op);
+                    Object output = e.getOutput();
+                    ctx.getCallback().onOperationExit(output);
```
Comment (nelsonsilva) [inline modules/platform/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/impl/OperationChainCompiler.java:218]: Good catch 👍  perhaps worth checking if anyone's using `ExitException`

**Review by nelsonsilva — APPROVED** [review]

File: modules/platform/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/jaxrs/adapters/AsyncOperationAdapter.java [hunk 3]
```diff
@@ -195,12 +199,12 @@ public OperationException onError(OperationException error) {
...
                         opCtx.setCoreSession(s);
                         service.run(opCtx, opId, xreq.getParams());
-                    } catch (OperationException e) {
+                    } catch (RuntimeException | OperationException e) {
                         setError(executionId, e);
                     } finally {
                         LoginComponent.popPrincipal();
                     }
-	         });
+             });
```
[the remark this answers is not in the document]
Comment (efge) [reply modules/platform/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/jaxrs/adapters/AsyncOperationAdapter.java:207]: There was a tab but the indent was indeed not fixed correctly.

File: modules/platform/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/jaxrs/adapters/AsyncOperationAdapter.java [hunk 4]
```diff
@@ -305,18 +304,36 @@ protected void enterMethod(String executionId, InvokableMethod method) {
...
     protected void setError(String executionId, Throwable t) {
-        setError(executionId, ExceptionHelper.unwrapException(t).getMessage());
-    }
-
-    /**
-     * @deprecated since 11.1, because unused
-     */
-    @Deprecated(since = "11.1")
-    protected void setError(String executionId, String error) {
+        // find custom status from any NuxeoException or RestOperationException
```
[the remark this answers is not in the document]
Comment (efge) [reply modules/platform/nuxeo-automation/nuxeo-automation-server/src/main/java/org/nuxeo/ecm/automation/server/jaxrs/adapters/AsyncOperationAdapter.java:307]: The eventual goal is to have everything derive from NuxeoException and use its status. But that's for another refactoring.

File: modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoRequestControllerFilter.java (hunk 1 above)
Comment (efge) [reply modules/platform/nuxeo-platform-web-common/src/main/java/org/nuxeo/ecm/platform/web/common/requestcontroller/filter/NuxeoRequestControllerFilter.java:142]: In order for the exception filter to be able to send a custom JSON with the exception, the existing output has to be cleared. And it has to be done here where we still have a buffered response that can be wiped out, before we give back control to the higher filter that doesn't have any buffering.

File: modules/platform/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/impl/OperationChainCompiler.java (hunk 2 above)
Comment (efge) [reply modules/platform/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/impl/OperationChainCompiler.java:218]: It seems that nobody knows about this feature.

**Comment by efge** [comment]
Force push to fix Locale management in test and in default exception filter.

**Comment by efge** [comment]
Force push to fix TestBulkRunAction that now needs more detailed error messages.

**Review by ataillefer — APPROVED** [review]

**Review by kevinleturc — APPROVED** [review]
